backfill/2022-08-15-clients-first-seen/backfill.py

```diff
@@ -0,0 +1,153 @@
+# Based on https://github.com/mozilla/bigquery-etl/blob/main/sql/moz-fx-data-shared-prod/telemetry_derived/clients_first_seen_v1/init.sql
+
+from argparse import ArgumentParser
+from multiprocessing.pool import ThreadPool
+from functools import partial
+
+from google.cloud import bigquery
+
+CREATE_TABLE_QUERY = """
+    CREATE OR REPLACE TABLE
+      {dataset}.{table}
+    PARTITION BY
+        first_seen_date
+    CLUSTER BY
+        normalized_channel,
+        sample_id
+    AS
+    SELECT
+        CAST(NULL AS DATE) AS first_seen_date,
+        CAST(NULL AS DATE) AS second_seen_date,
+        cd.* EXCEPT (submission_date)
+    FROM
+        `moz-fx-data-shared-prod.telemetry_derived.clients_daily_v6` AS cd
+    WHERE
+        FALSE;
+"""
+
+PARTITON_QUERY = """
+  CREATE OR REPLACE TABLE `moz-fx-data-shared-prod.tmp.clients_first_seen_dates_{sample_id}`
+  PARTITION BY
+    first_seen_date
+  AS
+  WITH base AS (
+    SELECT
+      client_id,
+      ARRAY_AGG(submission_date ORDER BY submission_date) AS dates_seen,
+    FROM
+      `moz-fx-data-shared-prod.telemetry_derived.clients_daily_v6`
+    WHERE
+      submission_date >= DATE('2010-01-01')
```

Comment:
Why this date? Is it because we need to scan the full table to get the first_seen_date of clients that got the wrong first_seen_date in 2022-05-11? Perhaps we could limit the scan to to the client_ids on or after this date?

Reply:
And below at line 66

Reply:
According to  Anna, clients_first_seen_v1 needs to be backfilled from the beginning: [https://mozilla.slack.com/archives/C03T47LEU4V/p1660547035708419](https://mozilla.slack.com/archives/C03T47LEU4V/p1660547035708419). But downstream tables need to be backfilled from May 11

Reply:
So we basically need to scan from the beginning to determine the very first entry of each client in clients_daily (which is the first seen date), which in theory could be '2010-01-01'.

Reply:
Thanks for confirming. What about using a subquery of client_ids affected? (although that would probably make the query not usable for general backfills) .
Just trying to understand if there are ways to not backfill all the data but only what is required.

Reply:
We could have skipped client_id for sample_ids that are not affected, but from what I can tell only 17 sample_ids got backfilled correctly, so to avoid missing some or skipping the wrong ones we might as well just run the backfill for all of them.